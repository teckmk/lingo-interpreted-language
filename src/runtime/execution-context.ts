export type StackFrame = {
  functionName: string
  filename: string
  position: {
    start: { line: number; column: number }
    end: { line: number; column: number }
  }
}

export class ExecutionContext {
  private callStack: StackFrame[]

  constructor() {
    this.callStack = []
  }

  public pushCallStack(frame: StackFrame) {
    this.callStack.push(frame)
  }

  public popCallStack() {
    this.callStack.pop()
  }

  public getCallStack(): StackFrame[] {
    return this.callStack.reverse()
  }

  static formatStackTrace(stack: StackFrame[], maxFrames = 10): string {
    const trimmedStack = stack.slice(0, maxFrames) // Limit frames to avoid excessive output
    let stackTrace = "Stack Trace:\n"
    for (const frame of trimmedStack) {
      // const shortFile = frame.filename.replace(/^.*[\\/]/, "") // Extract filename from path
      const shortFile = frame.filename
      stackTrace += `  at ${frame.functionName || "<anonymous>"} (${shortFile}:${frame.position.start.line}:${frame.position.start.column})\n`
    }

    if (stack.length > maxFrames) {
      stackTrace += `  ... (${stack.length - maxFrames} more frames)\n`
    }

    return stackTrace
  }
}
