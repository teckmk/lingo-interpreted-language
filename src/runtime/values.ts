import { Stmt } from "../frontend/2-ast"
import Environment from "./environment"

export type ValueType =
  | "null"
  | "number"
  | "string"
  | "boolean"
  | "object"
  | "nativefn"
  | "function"
  | "return"
  | "paramter"
  | "conditional"
  | "whileloop"
  | "array"
  | "dynamic"
  | "docs"

export interface RuntimeVal {
  type: ValueType
  returned: boolean
}

export interface NullVal extends RuntimeVal {
  type: "null"
  value: null
}
export interface BooleanVal extends RuntimeVal {
  type: "boolean"
  value: boolean
}

export interface NumberVal extends RuntimeVal {
  type: "number"
  value: number
}

export interface StringVal extends RuntimeVal {
  type: "string"
  value: string
}

export interface ObjectVal extends RuntimeVal {
  type: "object"
  properties: Map<string, RuntimeVal>
}

export type FunctionCall = (args: RuntimeVal[], env: Environment) => RuntimeVal

export interface NativeFnVal extends RuntimeVal {
  type: "nativefn"
  call: FunctionCall
}

export interface ParamVal extends RuntimeVal {
  type: "paramter"
  name: string
  valueType?: ValueType
  default: RuntimeVal // to assign a default value
}

export interface FunctionVal extends RuntimeVal {
  type: "function"
  name: string
  parameters: ParamVal[]
  declarationEnv: Environment
  body: Stmt[]
}

export interface ReturnVal extends RuntimeVal {
  type: "return"
  value: RuntimeVal
}

export interface ConditionalVal extends RuntimeVal {
  type: "conditional"
  check: RuntimeVal
  body: Stmt[]
  childChecks?: ConditionalVal[]
  else?: Stmt[]
}

export interface WhileLoopVal extends RuntimeVal {
  type: "whileloop"
  check: RuntimeVal
  body: Stmt[]
}

export interface ArrayVal extends RuntimeVal {
  type: "array"
  elements: RuntimeVal[]
}

export interface DocCommentVal extends RuntimeVal {
  type: "docs"
}
