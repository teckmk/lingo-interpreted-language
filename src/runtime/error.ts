import { ExecutionContext } from "./execution-context"

export class RuntimeError extends Error {
  context?: ExecutionContext

  constructor(context: ExecutionContext, message: string) {
    super(message)
    this.name = "RuntimeError"
    this.context = context
  }
}
