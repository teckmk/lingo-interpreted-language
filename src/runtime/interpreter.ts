import Environment from "./environment"
import { NumberVal, RuntimeVal, StringVal } from "./values"
import {
  ArrayLiteral,
  AssignmentExpr,
  BinaryExpr,
  CallExpr,
  DocComment,
  FunctionDeclaration,
  Identifier,
  IfElseStatement,
  MemberExpr,
  MultiVarDeclaration,
  NumericLiteral,
  ObjectLiteral,
  Program,
  ReturnStatement,
  Stmt,
  StringLiteral,
  VarDeclaration,
  WhileStatement,
} from "../frontend/2-ast"

import {
  eval_fn_declaration,
  eval_if_else_statement,
  eval_multi_var_declaration,
  eval_program,
  eval_return_statement,
  eval_var_declaration,
  eval_while_statement,
} from "./eval/statements"
import {
  eval_array_expr,
  eval_assignment,
  eval_binary_expr,
  eval_call_expr,
  eval_comment_expr,
  eval_identifier,
  eval_member_expr,
  eval_object_expr,
  eval_string_literal,
} from "./eval/expression"

export function evaluate(astNode: Stmt, env: Environment): RuntimeVal {
  switch (astNode.kind) {
    case "NumericLiteral":
      return {
        value: (astNode as NumericLiteral).value,
        type: "number",
      } as NumberVal
    case "StringLiteral":
      return eval_string_literal(astNode as StringLiteral, env) as StringVal
    case "BinaryExpr":
      return eval_binary_expr(astNode as BinaryExpr, env)
    case "AssignmentExpr":
      return eval_assignment(astNode as AssignmentExpr, env)
    case "Identifier":
      return eval_identifier(astNode as Identifier, env)
    case "ObjectLiteral":
      return eval_object_expr(astNode as ObjectLiteral, env)
    case "ArrayLiteral":
      return eval_array_expr(astNode as ArrayLiteral, env)
    case "CallExpr":
      return eval_call_expr(astNode as CallExpr, env)
    case "Program":
      return eval_program(astNode as Program, env)
    case "VarDeclaration":
      return eval_var_declaration(astNode as VarDeclaration, env)
    case "MultiVarDeclaration":
      return eval_multi_var_declaration(astNode as MultiVarDeclaration, env)
    case "FunctionDeclaration":
      return eval_fn_declaration(astNode as FunctionDeclaration, env)
    case "ReturnStatement":
      return eval_return_statement(astNode as ReturnStatement, env)
    case "IfElseStatement":
      return eval_if_else_statement(astNode as IfElseStatement, env)
    case "WhileStatement":
      return eval_while_statement(astNode as WhileStatement, env)
    case "MemberExpr":
      return eval_member_expr(astNode as MemberExpr, env)
    case "DocComment":
      return eval_comment_expr(astNode as DocComment, env)
    default:
      console.error("This AST can't be interpreted! For now atleast!", astNode)
      process.exit(0)
  }
}
