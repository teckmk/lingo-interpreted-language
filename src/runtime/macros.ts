import { BooleanVal, FunctionCall, NativeFnVal, NullVal, NumberVal, PlaceholderVal } from "./values"
import { PrimitiveTypeVal } from "./values.types"

export function MK_NUMBER(n = 0) {
  return { type: "number", value: n } as NumberVal
}

export function MK_NULL() {
  return { type: "null", value: null } as NullVal
}

export function MK_PLACEHOLDER() {
  return { type: "placeholder" } as PlaceholderVal
}

export function MK_BOOL(b: boolean) {
  return { type: "boolean", value: b } as BooleanVal
}

export function MK_NATIVE_FN(call: FunctionCall) {
  return { type: "nativefn", call } as NativeFnVal
}

export function MK_PRIM_TYPE(name: "number" | "bool" | "string" | "dynamic"): PrimitiveTypeVal {
  return {
    type: "type",
    typeName: name,
    typeKind: "primitive",
    primitiveType: name,
    returned: false, // To satify TS
  }
}
