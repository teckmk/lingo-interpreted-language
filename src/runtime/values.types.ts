import { RuntimeVal } from "./values"

export interface TypeVal extends RuntimeVal {
  type: "type"
  typeKind: string
  typeName?: string
  isNominal?: boolean // Track if this is a nominal type (created with 'type' keyword)
}

export interface PrimitiveTypeVal extends TypeVal {
  typeKind: "primitive"
  primitiveType: "string" | "number" | "bool" | "dynamic"
}

export interface StructTypeVal extends TypeVal {
  typeKind: "struct"
  members: Record<string, TypeVal>
  optional: Record<string, boolean> // Track which members are optional
}

export interface AliasTypeVal extends TypeVal {
  typeKind: "alias"
  aliasTo: TypeVal
}

export interface TypeParameterVal extends TypeVal {
  type: "type"
  typeKind: "typeParameter"
  name: string
  constraint?: TypeVal // i.e. T: number
}

export interface GenericTypeVal extends TypeVal {
  typeKind: "generic"
  baseType: TypeVal
  parameters: TypeParameterVal[]
}

export interface UnionTypeVal extends TypeVal {
  typeKind: "union"
  unionTypes: TypeVal[]
}

export interface ArrayTypeVal extends TypeVal {
  typeKind: "array"
  elementType: TypeVal
}
