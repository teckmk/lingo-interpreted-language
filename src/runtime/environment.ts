import { VarModifier } from "../frontend/ast"
import { ExecutionContext } from "./execution-context"
import { RuntimeError } from "./error"
import { ContractFulfillmentVal, FunctionVal, RuntimeVal, ValueType } from "./values"
import {
  TypeVal,
  PrimitiveTypeVal,
  AliasTypeVal,
  FunctionTypeVal,
  GetterTypeVal,
  ContractTypeVal,
} from "./values.types"
import { areTypesCompatible, createTypeArgMap, getRuntimeType, getTypeName } from "./type-checker"

import functions from "./built-ins/functions"
import variables from "./built-ins/variables"
import primitiveTypes from "./built-ins/types"

export default class Environment {
  private parent?: Environment
  private typeDefinitions: Map<string, RuntimeVal>
  private aliasDefinitions: Map<string, RuntimeVal>
  private variables: Map<string, RuntimeVal>
  private constants: Set<string>
  private finals: Set<string>
  private executionContext: ExecutionContext
  private variableTypes: Map<string, TypeVal>
  private contractFulfillments: Map<string, RuntimeVal>

  constructor(executionContext: ExecutionContext, parentEnv?: Environment) {
    this.parent = parentEnv
    this.executionContext = executionContext
    this.variables = new Map()
    this.constants = new Set()
    this.finals = new Set()
    this.typeDefinitions = new Map()
    this.aliasDefinitions = new Map()
    this.variableTypes = new Map()
    this.contractFulfillments = new Map()

    const global = Boolean(parentEnv) == false
    if (global) {
      // global variables
      for (const _var of variables) {
        this.declareVar(_var.name, _var.value, _var.modifier ?? "constant")
      }
      // global native functions
      for (const fn of functions) {
        this.declareVar(fn.name, fn.value, fn.modifier ?? "final")
      }

      // global primitive types
      for (const { name, value } of primitiveTypes) {
        this.declareType(name, value, true)
      }
    }
  }

  get context() {
    return this.executionContext
  }

  public declareType(name: string, value: RuntimeVal, isNominal = true): RuntimeVal {
    // Verify that we're dealing with a type
    if (value.type !== "type") {
      throw new RuntimeError(
        this.executionContext,
        `Cannot declare ${isNominal ? "type" : "alias"} ${name} with a non-type value.`,
      )
    }

    const alreadyDefined = this.typeDefinitions.has(name) || this.aliasDefinitions.has(name)

    if (isNominal) {
      if (alreadyDefined) {
        throw new RuntimeError(
          this.executionContext,
          `Cannot declare type ${name}. It is already defined.`,
        )
      }

      this.typeDefinitions.set(name, value)
      return value
    }

    if (alreadyDefined) {
      throw new RuntimeError(
        this.executionContext,
        `Cannot declare alias ${name}. It is already defined.`,
      )
    }

    // Prevent creating alias of an alias
    // Currently, our system wraps every type in AliasType, so we need to check if this alias contains another alias
    if (value.type == "type" && (value as TypeVal).typeKind == "alias") {
      const actualType = (value as AliasTypeVal).aliasTo
      if (actualType.type == "type" && actualType.typeKind == "alias")
        throw new RuntimeError(
          this.executionContext,
          `Cannot create an alias of another alias. Use the original type instead.`,
        )
    }

    this.aliasDefinitions.set(name, value)

    return value
  }

  public resolveType(name: string): Environment {
    if (this.typeDefinitions.has(name)) return this

    if (this.aliasDefinitions.has(name)) return this

    if (this.parent == undefined)
      throw new RuntimeError(this.executionContext, `Unable to resolve type ${name}`)

    return this.parent.resolveType(name)
  }

  public lookupType(name: string): RuntimeVal {
    const env = this.resolveType(name)

    let value
    if (env.aliasDefinitions.has(name)) {
      value = env.aliasDefinitions.get(name) as TypeVal
      value.isNominal = false
    } else if (env.typeDefinitions.has(name)) {
      value = env.typeDefinitions.get(name)
    }
    return value as RuntimeVal
  }

  private assertType(type: TypeVal | ValueType, value: RuntimeVal): RuntimeVal {
    // Handle primitive type strings for backward compatibility
    if (typeof type === "string") {
      // Dynamic type accepts any value
      if (type === "dynamic") {
        return { ...value, type: value.type }
      }

      // Convert string type to PrimitiveTypeVal
      const primitiveType: PrimitiveTypeVal = {
        type: "type",
        typeKind: "primitive",
        primitiveType: type as any,
        returned: false,
      }

      return this.assertType(primitiveType, value)
    }

    // If the value is dynamic, it can be assigned to any type
    if (value.type === "dynamic") {
      return { ...value, type: value.type }
    }

    // Get the runtime type of the value
    const valueType = getRuntimeType(value)

    // Check if the types are compatible
    const [isCompatible] = areTypesCompatible(valueType, type)

    if (!isCompatible) {
      const sourceTypeName =
        valueType.typeKind === "primitive"
          ? (valueType as PrimitiveTypeVal).primitiveType
          : valueType.typeKind

      const targetTypeName =
        type.typeKind === "primitive" ? (type as PrimitiveTypeVal).primitiveType : type.typeKind

      throw new RuntimeError(
        this.executionContext,
        `Type error: Cannot assign a value of type ${sourceTypeName} to a variable of type ${targetTypeName}`,
      )
    }

    return value
  }

  public declareVar(
    varname: string,
    value: RuntimeVal,
    modifier: VarModifier,
    type?: TypeVal | ValueType,
  ): RuntimeVal {
    if (this.variables.has(varname))
      throw new RuntimeError(
        this.executionContext,
        `Cannot declare variable ${varname}. As it already is defined.`,
      )

    switch (modifier) {
      case "final":
        this.finals.add(varname)
        break
      case "constant":
        this.constants.add(varname)
        break
    }

    let valueWithType = value

    if (type) {
      valueWithType = this.assertType(type, value)

      // Store the variable's type for future assignments
      if (typeof type !== "string") {
        this.variableTypes.set(varname, type)
      } else {
        // Convert string type to TypeVal
        const primitiveType: PrimitiveTypeVal = {
          type: "type",
          typeKind: "primitive",
          primitiveType: type as any,
          returned: false,
        }
        this.variableTypes.set(varname, primitiveType)
      }
    } else {
      // Infer the type from the value
      const inferredType = getRuntimeType(value)
      this.variableTypes.set(varname, inferredType)
    }

    this.variables.set(varname, valueWithType)

    return valueWithType
  }

  public assignVar(varname: string, value: RuntimeVal): RuntimeVal {
    const env = this.resolve(varname)

    // Can't assign to a constant or final variable
    if (this.finals.has(varname))
      throw new RuntimeError(
        this.executionContext,
        `Cannot assign to a final variable "${varname}"`,
      )
    else if (this.constants.has(varname))
      throw new RuntimeError(
        this.executionContext,
        `Cannot assign to a constant variable "${varname}"`,
      )

    const prevVal = env.lookupVar(varname)

    // Get the variable's declared type
    const varType = env.variableTypes.get(varname)

    let valueWithType = value

    if (varType) {
      valueWithType = env.assertType(varType, value)
    } else {
      // If no type was explicitly declared, ensure the new value matches the previous value's type
      const prevType = getRuntimeType(prevVal)
      valueWithType = env.assertType(prevType, value)
    }

    env.variables.set(varname, valueWithType)

    return valueWithType
  }

  public resolve(varname: string): Environment {
    if (this.variables.has(varname)) return this

    if (this.parent == undefined)
      throw new RuntimeError(
        this.executionContext,
        `Cannot resolve ${varname} as it does not exist.`,
      )

    return this.parent.resolve(varname)
  }

  public lookupVar(varname: string): RuntimeVal {
    const env = this.resolve(varname)
    return env.variables.get(varname) as RuntimeVal
  }

  public fulfillContract(
    structName: string,
    contractName: string,
    implementations: FunctionVal[],
    typeArgs: RuntimeVal[] = [],
  ): RuntimeVal {
    // Check if the struct exists
    if (!this.typeDefinitions.has(structName)) {
      // Check if it exists in a parent environment
      const structExists = this.parent && this.lookupType(structName)
      if (!structExists) {
        throw new RuntimeError(
          this.executionContext,
          `Cannot implement contract for ${structName}. The struct does not exist.`,
        )
      }
    }

    // Check if the contract exists
    if (!this.typeDefinitions.has(contractName)) {
      // Check if it exists in a parent environment
      const contractExists = this.parent && this.lookupType(contractName)
      if (!contractExists) {
        throw new RuntimeError(
          this.executionContext,
          `Cannot implement contract ${contractName}. The contract does not exist.`,
        )
      }
    }

    // Get the contract definition
    const contractType = this.lookupType(contractName) as ContractTypeVal
    const isDefaultContract = contractName === structName
    if (contractType.typeKind !== "contract" && !isDefaultContract) {
      throw new RuntimeError(
        this.executionContext,
        `Cannot implement ${contractName} because it is not a contract.`,
      )
    }

    // Enforce the orphan rule:
    // The implementation is valid only if either the contract or the struct is defined in this environment
    const isStructLocal = this.typeDefinitions.has(structName)
    const isContractLocal = this.typeDefinitions.has(contractName)

    if (!isStructLocal && !isContractLocal) {
      throw new RuntimeError(
        this.executionContext,
        `Orphan rule violation: Cannot implement contract ${contractName} for struct ${structName} because neither is defined in the current scope.`,
      )
    }

    // Generate a unique key for this implementation
    const fulfillmentKey = `${structName}:${contractName}`

    // Check if this contract is already implemented for this struct
    // We need to check the entire hierarchy to prevent duplicate implementations
    if (this.hasContractImplementation(structName, contractName)) {
      throw new RuntimeError(
        this.executionContext,
        `Contract ${contractName} is already implemented for ${structName}.`,
      )
    }

    // Verify that all required methods are implemented
    const contractMethods = Object.keys(contractType.members)
    const implementedMethods = implementations.map((fn) => fn.signature.name)

    // Check if all required methods are implemented
    if (!isDefaultContract) {
      for (const methodName of contractMethods) {
        if (!implementedMethods.includes(methodName)) {
          throw new RuntimeError(
            this.executionContext,
            `Implementation of contract ${contractName} is missing method ${methodName}, for struct ${structName}.`,
          )
        }

        // Verify method signatures match the contract
        const contractMethod = contractType.members[methodName] as FunctionTypeVal | GetterTypeVal
        // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
        const implementation = implementations.find((fn) => fn.signature.name === methodName)!

        // Check that the implementation's signature matches the contract's signature
        let argsMap
        const providedImpl = getRuntimeType(implementation)
        if (contractType.parameters) {
          argsMap = createTypeArgMap(contractType.parameters, typeArgs as TypeVal[])
        }
        const [isCompatible] = areTypesCompatible(providedImpl, contractMethod, argsMap)
        if (!isCompatible) {
          throw new RuntimeError(
            this.executionContext,
            `Method ${getTypeName(providedImpl)} of struct ${structName}, does not satisfy signature ${contractName}.${getTypeName(contractMethod)}.`,
          )
        }
      }
    }

    // Create the contract fulfillment value
    const fulfillment: ContractFulfillmentVal = {
      type: "fulfillment",
      contractName,
      for: structName,
      args: typeArgs,
      body: implementations,
      returned: false,
    }

    // Store the fulfillment
    this.contractFulfillments.set(fulfillmentKey, fulfillment)

    return fulfillment
  }

  public hasContractImplementation(structName: string, contractName: string): boolean {
    if (this.contractFulfillments.has(`${structName}:${contractName}`)) {
      return true
    }

    // Check parent environments
    return this.parent ? this.parent.hasContractImplementation(structName, contractName) : false
  }

  public getContractImplementation(
    structName: string,
    contractName: string,
  ): ContractFulfillmentVal | null {
    if (this.contractFulfillments.has(`${structName}:${contractName}`)) {
      return this.contractFulfillments.get(
        `${structName}:${contractName}`,
      ) as ContractFulfillmentVal
    }

    // Look in parent environments
    return this.parent ? this.parent.getContractImplementation(structName, contractName) : null
  }

  public doesImplementContract(structName: string, contractName: string): boolean {
    return this.hasContractImplementation(structName, contractName)
  }

  toJSON(): string {
    return "{}"
  }

  toString(): string {
    return "Environment"
  }
}
