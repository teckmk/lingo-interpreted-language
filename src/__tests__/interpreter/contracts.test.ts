import { interpret } from "../../runtime/interpreter";

describe("Interpreter - Contracts", () => {
  it("should interpret contracts", () => {
    const code = `
        type Point contract {
            fn setPoint(x:number, y:number) -> void
            get x -> number
            get y -> number
        }`;
    const runtimeVal = interpret("test", code);
    expect(runtimeVal).toMatchSnapshot();
  });

  it("should interpret contract fullfillment", () => {
    const code = `
        type Point contract {
            fn setPoint(x:number, y:number) -> void
            get x -> number
            get y -> number
        }

        type Point3D struct {
            x: number
            y: number
            z: number
        }

        fulfill Point for Point3D {
            fn setPoint(self, x:number, y:number) -> void {
                self.x = x
                self.y = y
            }
            get x(self) -> number {
                self.x
            }
            get y(self) -> number {
                self.y
            }
        }
`;
    const runtimeVal = interpret("test", code);
    expect(runtimeVal).toMatchSnapshot();
  });

  it("should not interpret contract fullfillment with wrong function signature", () => {
    const code = `
        type Point contract {
            fn setPoint(x:number, y:number) -> void
            get x -> number
        }
        type Point3D struct {
            x: number
            y: number
            z: number
        }
        fulfill Point for Point3D {
            fn setPoint(self, x:number) -> void {
                self.x = x
            }
            get x(self) -> number {
                self.x
            }
        }`;
    const runtimeVal = interpret("test", code);
    expect(runtimeVal).toEqual("Method setPoint(number) -> void of struct Point3D, does not satisfy signature Point.setPoint(number, number) -> void.\n");
  });

  it("should not interpret contract fullfillment with missing implementation", () => {
    const code = `
        type Point contract {
            fn setPoint(x:number, y:number) -> void
            get x -> number
        }
        type Point3D struct {
            x: number
            y: number
            z: number
        }
        fulfill Point for Point3D {
            fn setPoint(self, x:number, y:number) -> void {
                self.x = x
            }
          
        }`;
    const runtimeVal = interpret("test", code);
    expect(runtimeVal).toEqual("Implementation of contract Point is missing method x, for struct Point3D.\n");
  });
});
