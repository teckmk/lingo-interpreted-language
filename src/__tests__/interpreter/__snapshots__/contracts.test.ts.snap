// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Interpreter - Contracts should interpret contract fullfillment 1`] = `
{
  "args": [],
  "body": [
    {
      "body": [
        {
          "assigne": {
            "computed": false,
            "kind": "MemberExpr",
            "object": {
              "kind": "SelfKeyword",
              "symbol": {
                "position": {
                  "end": {
                    "column": 21,
                    "line": 16,
                  },
                  "start": {
                    "column": 17,
                    "line": 16,
                  },
                },
                "value": "self",
              },
            },
            "property": {
              "kind": "Identifier",
              "symbol": {
                "position": {
                  "end": {
                    "column": 23,
                    "line": 16,
                  },
                  "start": {
                    "column": 22,
                    "line": 16,
                  },
                },
                "value": "x",
              },
            },
          },
          "kind": "AssignmentExpr",
          "value": {
            "kind": "Identifier",
            "symbol": {
              "position": {
                "end": {
                  "column": 27,
                  "line": 16,
                },
                "start": {
                  "column": 26,
                  "line": 16,
                },
              },
              "value": "x",
            },
          },
        },
        {
          "assigne": {
            "computed": false,
            "kind": "MemberExpr",
            "object": {
              "kind": "SelfKeyword",
              "symbol": {
                "position": {
                  "end": {
                    "column": 21,
                    "line": 17,
                  },
                  "start": {
                    "column": 17,
                    "line": 17,
                  },
                },
                "value": "self",
              },
            },
            "property": {
              "kind": "Identifier",
              "symbol": {
                "position": {
                  "end": {
                    "column": 23,
                    "line": 17,
                  },
                  "start": {
                    "column": 22,
                    "line": 17,
                  },
                },
                "value": "y",
              },
            },
          },
          "kind": "AssignmentExpr",
          "value": {
            "kind": "Identifier",
            "symbol": {
              "position": {
                "end": {
                  "column": 27,
                  "line": 17,
                },
                "start": {
                  "column": 26,
                  "line": 17,
                },
              },
              "value": "y",
            },
          },
        },
      ],
      "declarationEnv": "{}",
      "returned": false,
      "signature": {
        "meta": {
          "isAsync": false,
          "isConstructor": false,
          "isGetter": false,
          "isMethod": true,
          "isStatic": false,
        },
        "name": "setPoint",
        "parameters": [
          {
            "default": {
              "type": "null",
              "value": null,
            },
            "name": "self",
            "returned": false,
            "type": "paramter",
            "valueType": undefined,
          },
          {
            "default": {
              "type": "null",
              "value": null,
            },
            "name": "x",
            "returned": false,
            "type": "paramter",
            "valueType": {
              "primitiveType": "number",
              "type": "type",
              "typeKind": "primitive",
            },
          },
          {
            "default": {
              "type": "null",
              "value": null,
            },
            "name": "y",
            "returned": false,
            "type": "paramter",
            "valueType": {
              "primitiveType": "number",
              "type": "type",
              "typeKind": "primitive",
            },
          },
        ],
        "returnType": [
          {
            "primitiveType": "void",
            "type": "type",
            "typeKind": "primitive",
          },
        ],
      },
      "type": "function",
    },
    {
      "body": [
        {
          "computed": false,
          "kind": "MemberExpr",
          "object": {
            "kind": "SelfKeyword",
            "symbol": {
              "position": {
                "end": {
                  "column": 21,
                  "line": 20,
                },
                "start": {
                  "column": 17,
                  "line": 20,
                },
              },
              "value": "self",
            },
          },
          "property": {
            "kind": "Identifier",
            "symbol": {
              "position": {
                "end": {
                  "column": 23,
                  "line": 20,
                },
                "start": {
                  "column": 22,
                  "line": 20,
                },
              },
              "value": "x",
            },
          },
        },
      ],
      "declarationEnv": "{}",
      "returned": false,
      "signature": {
        "meta": {
          "isAsync": false,
          "isConstructor": false,
          "isGetter": true,
          "isMethod": true,
          "isStatic": false,
        },
        "name": "x",
        "parameters": [
          {
            "default": {
              "type": "null",
              "value": null,
            },
            "name": "self",
            "returned": false,
            "type": "paramter",
            "valueType": undefined,
          },
        ],
        "returnType": [
          {
            "primitiveType": "number",
            "type": "type",
            "typeKind": "primitive",
          },
        ],
      },
      "type": "function",
    },
    {
      "body": [
        {
          "computed": false,
          "kind": "MemberExpr",
          "object": {
            "kind": "SelfKeyword",
            "symbol": {
              "position": {
                "end": {
                  "column": 21,
                  "line": 23,
                },
                "start": {
                  "column": 17,
                  "line": 23,
                },
              },
              "value": "self",
            },
          },
          "property": {
            "kind": "Identifier",
            "symbol": {
              "position": {
                "end": {
                  "column": 23,
                  "line": 23,
                },
                "start": {
                  "column": 22,
                  "line": 23,
                },
              },
              "value": "y",
            },
          },
        },
      ],
      "declarationEnv": "{}",
      "returned": false,
      "signature": {
        "meta": {
          "isAsync": false,
          "isConstructor": false,
          "isGetter": true,
          "isMethod": true,
          "isStatic": false,
        },
        "name": "y",
        "parameters": [
          {
            "default": {
              "type": "null",
              "value": null,
            },
            "name": "self",
            "returned": false,
            "type": "paramter",
            "valueType": undefined,
          },
        ],
        "returnType": [
          {
            "primitiveType": "number",
            "type": "type",
            "typeKind": "primitive",
          },
        ],
      },
      "type": "function",
    },
  ],
  "contractName": "Point",
  "for": "Point3D",
  "returned": false,
  "type": "fulfillment",
}
`;

exports[`Interpreter - Contracts should interpret contracts 1`] = `
{
  "isNominal": true,
  "members": {
    "setPoint": {
      "returned": false,
      "signature": {
        "meta": {
          "isAsync": false,
          "isConstructor": false,
          "isGetter": false,
          "isMethod": false,
          "isStatic": false,
        },
        "name": "setPoint",
        "parameters": [
          {
            "default": {
              "type": "null",
              "value": null,
            },
            "name": "x",
            "returned": false,
            "type": "paramter",
            "valueType": {
              "primitiveType": "number",
              "type": "type",
              "typeKind": "primitive",
            },
          },
          {
            "default": {
              "type": "null",
              "value": null,
            },
            "name": "y",
            "returned": false,
            "type": "paramter",
            "valueType": {
              "primitiveType": "number",
              "type": "type",
              "typeKind": "primitive",
            },
          },
        ],
        "returnType": [
          {
            "primitiveType": "void",
            "type": "type",
            "typeKind": "primitive",
          },
        ],
      },
      "type": "type",
      "typeKind": "function",
    },
    "x": {
      "returned": false,
      "signature": {
        "meta": {
          "isAsync": false,
          "isConstructor": false,
          "isGetter": true,
          "isMethod": false,
          "isStatic": false,
        },
        "name": "x",
        "parameters": [],
        "returnType": [
          {
            "primitiveType": "number",
            "type": "type",
            "typeKind": "primitive",
          },
        ],
      },
      "type": "type",
      "typeKind": "getter",
    },
    "y": {
      "returned": false,
      "signature": {
        "meta": {
          "isAsync": false,
          "isConstructor": false,
          "isGetter": true,
          "isMethod": false,
          "isStatic": false,
        },
        "name": "y",
        "parameters": [],
        "returnType": [
          {
            "primitiveType": "number",
            "type": "type",
            "typeKind": "primitive",
          },
        ],
      },
      "type": "type",
      "typeKind": "getter",
    },
  },
  "returned": false,
  "type": "type",
  "typeKind": "contract",
  "typeName": "Point",
}
`;

exports[`Interpreter - Contracts should interpret default contract fullfillment 1`] = `
{
  "args": [],
  "body": [
    {
      "body": [
        {
          "assigne": {
            "computed": false,
            "kind": "MemberExpr",
            "object": {
              "kind": "SelfKeyword",
              "symbol": {
                "position": {
                  "end": {
                    "column": 21,
                    "line": 9,
                  },
                  "start": {
                    "column": 17,
                    "line": 9,
                  },
                },
                "value": "self",
              },
            },
            "property": {
              "kind": "Identifier",
              "symbol": {
                "position": {
                  "end": {
                    "column": 23,
                    "line": 9,
                  },
                  "start": {
                    "column": 22,
                    "line": 9,
                  },
                },
                "value": "x",
              },
            },
          },
          "kind": "AssignmentExpr",
          "value": {
            "kind": "Identifier",
            "symbol": {
              "position": {
                "end": {
                  "column": 27,
                  "line": 9,
                },
                "start": {
                  "column": 26,
                  "line": 9,
                },
              },
              "value": "x",
            },
          },
        },
        {
          "assigne": {
            "computed": false,
            "kind": "MemberExpr",
            "object": {
              "kind": "SelfKeyword",
              "symbol": {
                "position": {
                  "end": {
                    "column": 21,
                    "line": 10,
                  },
                  "start": {
                    "column": 17,
                    "line": 10,
                  },
                },
                "value": "self",
              },
            },
            "property": {
              "kind": "Identifier",
              "symbol": {
                "position": {
                  "end": {
                    "column": 23,
                    "line": 10,
                  },
                  "start": {
                    "column": 22,
                    "line": 10,
                  },
                },
                "value": "y",
              },
            },
          },
          "kind": "AssignmentExpr",
          "value": {
            "kind": "Identifier",
            "symbol": {
              "position": {
                "end": {
                  "column": 27,
                  "line": 10,
                },
                "start": {
                  "column": 26,
                  "line": 10,
                },
              },
              "value": "y",
            },
          },
        },
      ],
      "declarationEnv": "{}",
      "returned": false,
      "signature": {
        "meta": {
          "isAsync": false,
          "isConstructor": false,
          "isGetter": false,
          "isMethod": true,
          "isStatic": false,
        },
        "name": "setPoint",
        "parameters": [
          {
            "default": {
              "type": "null",
              "value": null,
            },
            "name": "self",
            "returned": false,
            "type": "paramter",
            "valueType": undefined,
          },
          {
            "default": {
              "type": "null",
              "value": null,
            },
            "name": "x",
            "returned": false,
            "type": "paramter",
            "valueType": {
              "primitiveType": "number",
              "type": "type",
              "typeKind": "primitive",
            },
          },
          {
            "default": {
              "type": "null",
              "value": null,
            },
            "name": "y",
            "returned": false,
            "type": "paramter",
            "valueType": {
              "primitiveType": "number",
              "type": "type",
              "typeKind": "primitive",
            },
          },
        ],
        "returnType": [
          {
            "primitiveType": "void",
            "type": "type",
            "typeKind": "primitive",
          },
        ],
      },
      "type": "function",
    },
  ],
  "contractName": "Point3D",
  "for": "Point3D",
  "returned": false,
  "type": "fulfillment",
}
`;
